### PR TITLE
New platform independent UART interface

### PR DESCRIPTION
**Pull request by katzfey**
### Solved Problem
There are multiple copies of UART configuration and use code that are not platform independent. The Qurt platform, in particular, is quite different in terms of UART usage. This is a proposal for a new platform independent UART API that hides all of the implementation details in platform specific files and provides a generic API for all platforms. The GPS, RC, and UART based ESC drivers will important beneficiaries of this new API.

### CODE REVIEW

File: src/drivers/gps/gps.cpp [hunk 1]
```diff
@@ -449,72 +460,74 @@ int GPS::callback(GPSCallbackType type, void *data1, int data2, void *user)
...
-			 * won't block even on a blocking device. But don't read immediately
-			 * by 1-2 bytes, wait for some more data to save expensive read() calls.
-			 * If we have all requested data available, read it without waiting.
-			 * If more bytes are available, we'll go back to poll() again.
-			 */
-			const unsigned character_count = 32; // minimum bytes that we want to read
-			unsigned baudrate = _baudrate == 0 ? 115200 : _baudrate;
-			const unsigned sleeptime = character_count * 1000000 / (baudrate / 10);
+// Qurt does not support poll for serial devices (nor external SPI devices for that matter)
+#if !defined(__PX4_QURT)
```
[the remark this answers is not in the document]
Comment (katzfey) [reply src/drivers/gps/gps.cpp]: Okay, I changed SPI to be a Linux only thing.

**Comment by PetervdPerk-NXP** [comment]
Its seems that on all non-Qurt targets the flash usage roughly increases by 2KB
Bloaty output px4-fmuv2
` +0.2% +2.10Ki    TOTAL`

@dagar you might want to update bloaty on the CI since on many boards it gets an integer overflow.

PX4 pretty much assumes it's using POSIX based UART and uses termios to change settings.
Wouldn't it be better to add a shim to platforms/qurt to implement the POSIX uart + termios for Qurt?

**Comment by dagar** [comment]
> @dagar you might want to update bloaty on the CI since on many boards it gets an integer overflow.

Thanks @PetervdPerk-NXP, I'm going to update the containers to 22.04 + latest tools soon.

> PX4 pretty much assumes it's using POSIX based UART and uses termios to change settings.
Wouldn't it be better to add a shim to platforms/qurt to implement the POSIX uart + termios for Qurt?

That was one idea I proposed initially, but after some discussion with @katzfey and my own pain with subtle differences in POSIX between NuttX/Linux/MacOS I actually thought this could be a nice opportunity to simplify/optimize the vast majority of our serial usage with an API that's easier to use, harder to misuse.

 - vast majority of serial usage is "raw" mode with that wall of termios + cfsetispeed configuration duplicated across most drivers
 - we don't actually use most of POSIX terminal I/O past initial config where we disable everything to get to raw mode
 - there are subtle differences between NuttX/Linux/MacOS with O_NONBLOCK, cfsetspeed, VTIME, VMIN, etc
 - opportunity to optimize the typical usage pattern where we want at least a certain number of bytes with minimal latency (NuttX doesn't actually implement VMIN/VTIME)
 - simpler setup and cleanup (RAII, no direct file descriptors, etc)
 - possibly a bit easier for future ports to Zephyr or even crazier ideas like a native windows build (not a major consideration)

The immediate goal in this PR was to get to an acceptable MVP to unblock QuRT progress and leave room to continue some of the other ideas in the future.

Happy to discuss further before we fully commit to anything.

File: src/drivers/gps/gps.cpp [hunk 2]
```diff
@@ -449,72 +469,64 @@ int GPS::callback(GPSCallbackType type, void *data1, int data2, void *user)
 
 int GPS::pollOrRead(uint8_t *buf, size_t buf_length, int timeout)
 {
+	int ret = 0;
+	const unsigned character_count = 32; // minimum bytes that we want to read
+	const int max_timeout = 50;
+	int _timeout = math::min(max_timeout, timeout);
```
[the remark this answers is not in the document]
Comment (katzfey) [reply src/drivers/gps/gps.cpp]: Done!

**Comment by SalimTerryLi** [comment]
I just come up with some other functionality requirements related to UART:

- Single-Wire mode
- Inverted mode
- WorkQueue integration? (mostly about `px4::serial_port_to_wq()` so that per-bus thread/task can be achieved)
- Possibility to do non-blocking IO with `FIONREAD` & `FIONSPACE` support instead of fire-until-timeout? (not sure if any driver requires that)

Just gave up reading the mavlink module for its size XD Not sure what requirement is needed for those complex, software throttling enabled modules.

The temporary increased code size will be reduced once drivers switching to new API for sure... Just noticed so many  redundant occurrences...

File: platforms/nuttx/src/px4/common/px4_layer.cmake [hunk 3]
```diff
@@ -3,6 +3,7 @@
 add_library(px4_layer
 		${KERNEL_SRCS}
 		cdc_acm_check.cpp
+		SerialImpl.cpp
```
Comment (dirksavage88) [inline platforms/nuttx/src/px4/common/px4_layer.cmake:6]: Add this also to the [px4_protected_layers.cmake](https://github.com/PX4/PX4-Autopilot/blob/42547a1319d0952d15ef845b8b8560eabf5cb7e5/platforms/nuttx/src/px4/common/px4_protected_layers.cmake#L15)

File: platforms/common/CMakeLists.txt [hunk 4]
```diff
@@ -51,9 +51,10 @@ add_library(px4_platform STATIC
 	px4_cli.cpp
 	shutdown.cpp
 	spi.cpp
+	Serial.cpp
 	${SRCS}
 )
-target_link_libraries(px4_platform prebuild_targets px4_work_queue)
+target_link_libraries(px4_platform prebuild_targets px4_work_queue px4_layer)
```
Comment (dirksavage88) [inline platforms/common/CMakeLists.txt:57]: I checked and this is causing the redefinition errors.

**Review by dirksavage88 — COMMENTED** [review]
@dagar is it necessary to split the driver into the common and nuttx (and POSIX?) layer or can it all be lumped at the px4 common layer? Serial.cpp needs to be linked with SerialImp.cpp to resolve linking errors it seems.

File: platforms/common/CMakeLists.txt (hunk 4 above)
Comment (katzfey) [reply platforms/common/CMakeLists.txt:57]: The whole point is that there is a common part of the driver (Serial.cpp) and a platform specific part of the driver (SerialImpl.cpp). Each platform has it's own platform specific version of SerialImpl.cpp.

Comment (katzfey) [reply platforms/common/CMakeLists.txt:57]: The redefinition errors are only occurring with the protected build, not the default build. So there is something about the linking for that type of build that requires a different approach.

Comment (katzfey) [reply platforms/common/CMakeLists.txt]: I was able to resolve the issue in the latest commit.

**Comment by katzfey** [comment]
@SalimTerryLi The idea with this PR is to get started with GPS only. So the new API has enough functionality for only what GPS needs and GPS is updated to use the new API. This is actually a pretty standard configuration so it should be usable by most other drivers. But as more drivers get updated to use the new API any missing additional functionality or settings that they need will need to be added to the Serial driver at that time. This will especially be true for the RC drivers as they have a bunch of oddball requirements.

File: platforms/nuttx/src/px4/common/SerialImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+		fds[0].fd = _serial_fd;
+		fds[0].events = POLLIN;
+
+		hrt_abstime remaining_time = timeout_us - hrt_elapsed_time(&start_time_us);
+
+		if (remaining_time <= 0) { break; }
+
+		int ret = poll(fds, sizeof(fds) / sizeof(fds[0]), remaining_time);
+
+		if (ret > 0) {
```
Comment (dagar) [inline platforms/nuttx/src/px4/common/SerialImpl.cpp:277]: Once even a single byte is available (but < character_count) isn't this just going to spin constantly? 
Was there a problem with the sleep that was here originally?

Comment (katzfey) [reply platforms/nuttx/src/px4/common/SerialImpl.cpp:277]: Oh, I didn't mean to change anything. Let me take a look.

Comment (katzfey) [reply platforms/nuttx/src/px4/common/SerialImpl.cpp:277]: I put them back in

**Review by dagar — APPROVED** [review]

**Comment by PetervdPerk-NXP** [comment]
Oh this went in?
I'm still missing the routines for TX/RX swap, TX/RX invert and Single-wire mode. Any roadmap to get this in? Kinda sucks to havell have 2 living UART subsystems.

**Comment by katzfey** [comment]
@PetervdPerk-NXP Yes, those will be coming in as well. The idea was to add enough functionality for the drivers being ported over. Initially this is just GPS where those other features are not used. But today I am working on PR 22917 which requires flush, Tx / Rx swap, and single wire. So those will be added in that PR. Next on the list is UART distance sensor and RC Library which should bring in all the rest of the needed UART functions.

**Comment by julianoes** [comment]
This unfortunately breaks Septentrino GPS with the warnings:

```
ERROR [SerialImpl] readAtLeast: Buffer not big enough to hold desired amount of read data
```

**Comment by julianoes** [comment]
Fix in https://github.com/PX4/PX4-Autopilot/pull/22936.

**Comment by kevinuav** [comment]
Hi,Eric. I wanna rebuild my code with your serial's API. I follow what you did over the gps's driver. However it encounters error while it compiles. What do I miss?
The error:/Users/kevinchen/PX4-Autopilot/src/drivers/gasengine/gasengine.hpp:139:2: error: 'Serial' does not name a type
  139 |  Serial     _uart {};     ///< UART interface to gaseng
      |  ^~~~~~
compilation terminated due to -Wfatal-errors.
截屏2024-07-12 19 24 49

**Comment by katzfey** [comment]
@kevinuav Did you "#include <px4_platform_common/Serial.hpp>"?

**Comment by kevinuav** [comment]
> @kevinuav Did you "#include <px4_platform_common/Serial.hpp>"?

Of course.

**Comment by kevinuav** [comment]
> @kevinuav Did you "#include <px4_platform_common/Serial.hpp>"?
Solved. I forgot the namespace. Sorry. And thank you.

**Comment by kevinuav** [comment]
Hi, Katzfey. I find a bug in you PX4's Serial API.I use the readAtLeast() to get the command reply from my uart device. The device will reply my command first. Then it sends back the messages accord to my command. 
Just like this: 
**I send: 1,APE,1\n
It sendback:1,APE,1\n message.......** 
However I find the strange thing is readAtLeast() can't send back the command"**1,APE,1\n**". Why? Is it designed to cut off the first characters I just sent? Or it is just a bug?

**Comment by katzfey** [comment]
@kevinuav Are you saying it works as expected when you don't use the Serial API, but when you switch to the Serial API you lose part of the response? What platform are you on? NuttX?

**Comment by kevinuav** [comment]
> @kevinuav Are you saying it works as expected when you don't use the Serial API, but when you switch to the Serial API you lose part of the response? What platform are you on? NuttX?

I have just checkout the latest firmware. I am using CUAV's V5-nano. The compile target is px4_fmu-v5_default.

**Comment by kevinuav** [comment]
> @kevinuav Are you saying it works as expected when you don't use the Serial API, but when you switch to the Serial API you lose part of the response? What platform are you on? NuttX?


I have just checkout the latest firmware.It should be the latest version of NuttX. I am using CUAV's V5-nano. The compile target is px4_fmu-v5_default.

**Comment by katzfey** [comment]
The code for nuttx readAtLeast is here: https://github.com/PX4/PX4-Autopilot/blob/36d89df0a722e2fca3f23e4d0ddd9da9fb5b7ecc/platforms/nuttx/src/px4/common/SerialImpl.cpp#L254. You can see that it does not look at the data. What code are you working with that is using the Serial API?

**Pull request merged**